### ISSUE TITLE
ModuleNotFoundError: No module named 'piblo'

### ISSUE THREAD

**Issue by danmcpherson** [jezs00/pycasso#48]
Thanks for the tool, looking forward to getting it running.

I think things installed correctly, however, I'm getting the following error when I try to run many functions (the API Key set, the review screen). 

Traceback (most recent call last):
  File "/home/danmc/pycasso/examples/review_screen.py", line 8, in <module>
    from piblo import config_wrapper
ModuleNotFoundError: No module named 'piblo'

I've tried making some changes to my PYTHONPATH without luck. I expect it's a pretty quick fix. It's on a fresh Raspberry Pi Zero W. 

Thanks again,
Daniel

**Comment by jezs00** [comment]
That would mean it's not finding the piblo package which should install in step 1. Have a go running `bash <(curl https://raw.githubusercontent.com/jezs00/pycasso/main/setup.sh)` again and using option ` 1 - Install/Upgrade pycasso (Full)` then close the installer.

Have a look at the terminal after closing for results of the package installs and note any errors regarding install of piblo package. Try running again, and if it's still not working post any errors you have here.

**Comment by jezs00** [comment]
Hmm, I might need to investigate further, it seems like some of the
requirements for piblo are restricting our ability to install properly. Can
you also try to update/install pycasso minimally from the installer? It
should be option 2 from memory, this should install piblo without stopping
for problems with requirements.

**Comment by BigDave365** [comment]
I have the same hardware and the same issue. Pycasso is not generating an error log (at least one I can find), and it does not show up in running services. If run the run.py script, it says that the module "webuiapi" is missing.

I also tried running on a Pi 3B and it all worked without a hitch....

**Comment by jezs00** [comment]
That is annoying, I think what has happened is some requirements have been
missed due to failures, but I can't see evidence of that in the pip log you
pasted. Perhaps there's a package that doesn't work well on this hardware
or we've added a package recently that is causing problems. I can probably
have a better look this weekend. Otherwise, if you guys can check out
requirements.txt and run 'sudo pip3 install' for each package in
requirements.txt, it should help:
- will show if one package in particular is giving a problem
- hopefully will handle packages you're missing

Let me know how you go.

**Comment by BigDave365** [comment]
I gave it a go. I had a few issues figuring out the syntax (not a Linux guy), but seem to have several packages missing. All seemed to install Ok save Stability-AI which threw an error involving Gripco (the last one in the log).

Here is the Log
dave@rpipictureframe:~/pycasso $ sudo pip3 install pillow
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Requirement already satisfied: pillow in /usr/local/lib/python3.9/dist-packages                                                                                                                      (9.4.0)
dave@rpipictureframe:~/pycasso $ sudo pip3 install keyring
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Collecting keyring
  Using cached https://www.piwheels.org/simple/keyring/keyring-23.13.1-py3-none-                                                                                                                     any.whl (37 kB)
Collecting jeepney>=0.4.2
  Downloading https://www.piwheels.org/simple/jeepney/jeepney-0.8.0-py3-none-any                                                                                                                     .whl (48 kB)
     |████████████████████████████████| 48 kB 205 kB/s
Collecting SecretStorage>=3.2
  Downloading https://www.piwheels.org/simple/secretstorage/SecretStorage-3.3.3-                                                                                                                     py3-none-any.whl (15 kB)
Collecting jaraco.classes
  Downloading https://www.piwheels.org/simple/jaraco-classes/jaraco.classes-3.2.                                                                                                                     3-py3-none-any.whl (6.0 kB)
Collecting importlib-metadata>=4.11.4
  Downloading https://www.piwheels.org/simple/importlib-metadata/importlib_metad                                                                                                                     ata-6.0.0-py3-none-any.whl (21 kB)
Collecting zipp>=0.5
  Downloading https://www.piwheels.org/simple/zipp/zipp-3.14.0-py3-none-any.whl                                                                                                                      (6.7 kB)
Requirement already satisfied: cryptography>=2.0 in /usr/lib/python3/dist-packag                                                                                                                     es (from SecretStorage>=3.2->keyring) (3.3.2)
Collecting more-itertools
  Downloading https://www.piwheels.org/simple/more-itertools/more_itertools-9.0.                                                                                                                     0-py3-none-any.whl (52 kB)
     |████████████████████████████████| 52 kB 147 kB/s
Installing collected packages: zipp, more-itertools, jeepney, SecretStorage, jar                                                                                                                     aco.classes, importlib-metadata, keyring
ERROR: pip's dependency resolver does not currently take into account all the pa                                                                                                                     ckages that are installed. This behaviour is the source of the following depende                                                                                                                     ncy conflicts.
piblo 1.2.0 requires mastodon.py, which is not installed.
piblo 1.2.0 requires openai@ git+https://github.com/openai/openai-python.git, wh                                                                                                                     ich is not installed.
piblo 1.2.0 requires stability-sdk@ git+https://github.com/Stability-AI/stabilit                                                                                                                     y-sdk.git, which is not installed.
piblo 1.2.0 requires webuiapi, which is not installed.
Successfully installed SecretStorage-3.3.3 importlib-metadata-6.0.0 jaraco.class                                                                                                                     es-3.2.3 jeepney-0.8.0 keyring-23.13.1 more-itertools-9.0.0 zipp-3.14.0
dave@rpipictureframe:~/pycasso $ sudo pip3 install https://github.com/robweber/o                                                                                                                     mni-epd.git
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Collecting https://github.com/robweber/omni-epd.git
  Downloading https://github.com/robweber/omni-epd.git
     - 247 kB 705 kB/s
  ERROR: Cannot unpack file /tmp/pip-unpack-3lnlmvlw/omni-epd.git (downloaded fr                                                                                                                     om /tmp/pip-req-build-cw6dqf_q, content-type: text/html; charset=utf-8); cannot                                                                                                                      detect archive format
ERROR: Cannot determine archive format of /tmp/pip-req-build-cw6dqf_q
dave@rpipictureframe:~/pycasso $ sudo pip3 install https://github.com/robweber/o                                                                                                                     mni-epd.git#egg=omni-epd
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Collecting omni-epd
  Downloading https://github.com/robweber/omni-epd.git
     - 243 kB 85 kB/s
  ERROR: Cannot unpack file /tmp/pip-unpack-cuquq6gy/omni-epd.git (downloaded fr                                                                                                                     om /tmp/pip-install-7ramk8_w/omni-epd_7c255aa69dc948e9a9179faec23c1e7a, content-                                                                                                                     type: text/html; charset=utf-8); cannot detect archive format
ERROR: Cannot determine archive format of /tmp/pip-install-7ramk8_w/omni-epd_7c2                                                                                                                     55aa69dc948e9a9179faec23c1e7a
dave@rpipictureframe:~/pycasso $ sudo pip3 install https://github.com/Stability-                                                                                                                     AI/stability-sdk.git
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Collecting https://github.com/Stability-AI/stability-sdk.git
  Downloading https://github.com/Stability-AI/stability-sdk.git
     \ 234 kB 585 kB/s
  ERROR: Cannot unpack file /tmp/pip-unpack-69wm2fvf/stability-sdk.git (download                                                                                                                     ed from /tmp/pip-req-build-z9hipfca, content-type: text/html; charset=utf-8); ca                                                                                                                     nnot detect archive format
ERROR: Cannot determine archive format of /tmp/pip-req-build-z9hipfca
dave@rpipictureframe:~/pycasso $ sudo pip3 install requests
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Requirement already satisfied: requests in /usr/lib/python3/dist-packages (2.25.                                                                                                                     1)
dave@rpipictureframe:~/pycasso $ sudo pip3 install numpy
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Requirement already satisfied: numpy in /usr/lib/python3/dist-packages (1.19.5)
dave@rpipictureframe:~/pycasso $ sudo pip3 install webuiapi
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Collecting webuiapi
  Using cached webuiapi-0.3.3-py3-none-any.whl (9.3 kB)
Requirement already satisfied: Pillow in /usr/local/lib/python3.9/dist-packages                                                                                                                      (from webuiapi) (9.4.0)
Requirement already satisfied: requests in /usr/lib/python3/dist-packages (from                                                                                                                      webuiapi) (2.25.1)
Installing collected packages: webuiapi
ERROR: pip's dependency resolver does not currently take into account all the pa                                                                                                                     ckages that are installed. This behaviour is the source of the following depende                                                                                                                     ncy conflicts.
piblo 1.2.0 requires mastodon.py, which is not installed.
piblo 1.2.0 requires openai@ git+https://github.com/openai/openai-python.git, wh                                                                                                                     ich is not installed.
piblo 1.2.0 requires stability-sdk@ git+https://github.com/Stability-AI/stabilit                                                                                                                     y-sdk.git, which is not installed.
Successfully installed webuiapi-0.3.3
dave@rpipictureframe:~/pycasso $ sudo pip3 install mastodon.py
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Collecting mastodon.py
  Using cached https://www.piwheels.org/simple/mastodon-py/Mastodon.py-1.8.0-py2                                                                                                                     .py3-none-any.whl (64 kB)
Collecting decorator>=4.0.0
  Downloading https://www.piwheels.org/simple/decorator/decorator-5.1.1-py3-none                                                                                                                     -any.whl (10 kB)
Requirement already satisfied: requests>=2.4.2 in /usr/lib/python3/dist-packages                                                                                                                      (from mastodon.py) (2.25.1)
Collecting python-magic
  Downloading https://www.piwheels.org/simple/python-magic/python_magic-0.4.27-p                                                                                                                     y2.py3-none-any.whl (15 kB)
Collecting blurhash>=1.1.4
  Downloading https://www.piwheels.org/simple/blurhash/blurhash-1.1.4-py2.py3-no                                                                                                                     ne-any.whl (5.3 kB)
Collecting python-dateutil
  Downloading https://www.piwheels.org/simple/python-dateutil/python_dateutil-2.                                                                                                                     8.2-py2.py3-none-any.whl (247 kB)
     |████████████████████████████████| 247 kB 281 kB/s
Requirement already satisfied: six in /usr/lib/python3/dist-packages (from masto                                                                                                                     don.py) (1.16.0)
Installing collected packages: python-magic, python-dateutil, decorator, blurhas                                                                                                                     h, mastodon.py
ERROR: pip's dependency resolver does not currently take into account all the pa                                                                                                                     ckages that are installed. This behaviour is the source of the following depende                                                                                                                     ncy conflicts.
piblo 1.2.0 requires openai@ git+https://github.com/openai/openai-python.git, wh                                                                                                                     ich is not installed.
piblo 1.2.0 requires stability-sdk@ git+https://github.com/Stability-AI/stabilit                                                                                                                     y-sdk.git, which is not installed.
Successfully installed blurhash-1.1.4 decorator-5.1.1 mastodon.py-1.8.0 python-d                                                                                                                     ateutil-2.8.2 python-magic-0.4.27
dave@rpipictureframe:~/pycasso $  git+https://github.com/openai/openai-python.gi                                                                                                                     t
-bash: git+https://github.com/openai/openai-python.git: No such file or director                                                                                                                     y
dave@rpipictureframe:~/pycasso $  git+https://github.com/Stability-AI/stability-                                                                                                                     sdk.git
-bash: git+https://github.com/Stability-AI/stability-sdk.git: No such file or di                                                                                                                     rectory
dave@rpipictureframe:~/pycasso $ sudo pip3 install https://github.com/Stability-                                                                                                                     AI/stability-sdk.git
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Collecting https://github.com/Stability-AI/stability-sdk.git
  Downloading https://github.com/Stability-AI/stability-sdk.git
     \ 232 kB 899 kB/s
  ERROR: Cannot unpack file /tmp/pip-unpack-ca5aa42i/stability-sdk.git (download                                                                                                                     ed from /tmp/pip-req-build-ckry5ogp, content-type: text/html; charset=utf-8); ca                                                                                                                     nnot detect archive format
ERROR: Cannot determine archive format of /tmp/pip-req-build-ckry5ogp
dave@rpipictureframe:~/pycasso $ sudo pip3 install "git+https://github.com/opena                                                                                                                     i/openai-python.git"
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Collecting git+https://github.com/openai/openai-python.git
  Cloning https://github.com/openai/openai-python.git to /tmp/pip-req-build-qvj8                                                                                                                     xo92
  Running command git clone -q https://github.com/openai/openai-python.git /tmp/                                                                                                                     pip-req-build-qvj8xo92
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
    Preparing wheel metadata ... done
Collecting aiohttp
  Using cached https://www.piwheels.org/simple/aiohttp/aiohttp-3.8.4-cp39-cp39-l                                                                                                                     inux_armv6l.whl (926 kB)
Collecting tqdm
  Using cached https://www.piwheels.org/simple/tqdm/tqdm-4.64.1-py2.py3-none-any                                                                                                                     .whl (78 kB)
Requirement already satisfied: requests>=2.20 in /usr/lib/python3/dist-packages                                                                                                                      (from openai==0.26.5) (2.25.1)
Collecting multidict<7.0,>=4.5
  Downloading https://www.piwheels.org/simple/multidict/multidict-6.0.4-cp39-cp3                                                                                                                     9-linux_armv6l.whl (109 kB)
     |████████████████████████████████| 109 kB 324 kB/s
Collecting attrs>=17.3.0
  Downloading https://www.piwheels.org/simple/attrs/attrs-22.2.0-py3-none-any.wh                                                                                                                     l (60 kB)
     |████████████████████████████████| 60 kB 241 kB/s
Collecting frozenlist>=1.1.1
  Downloading https://www.piwheels.org/simple/frozenlist/frozenlist-1.3.3-cp39-c                                                                                                                     p39-linux_armv6l.whl (136 kB)
     |████████████████████████████████| 136 kB 307 kB/s
Collecting aiosignal>=1.1.2
  Downloading https://www.piwheels.org/simple/aiosignal/aiosignal-1.3.1-py3-none                                                                                                                     -any.whl (7.6 kB)
Collecting yarl<2.0,>=1.0
  Downloading https://www.piwheels.org/simple/yarl/yarl-1.8.2-cp39-cp39-linux_ar                                                                                                                     mv6l.whl (205 kB)
     |████████████████████████████████| 205 kB 321 kB/s
Collecting charset-normalizer<4.0,>=2.0
  Downloading https://www.piwheels.org/simple/charset-normalizer/charset_normali                                                                                                                     zer-3.0.1-py3-none-any.whl (45 kB)
     |████████████████████████████████| 45 kB 182 kB/s
Collecting async-timeout<5.0,>=4.0.0a3
  Downloading https://www.piwheels.org/simple/async-timeout/async_timeout-4.0.2-                                                                                                                     py3-none-any.whl (5.8 kB)
Requirement already satisfied: idna>=2.0 in /usr/lib/python3/dist-packages (from                                                                                                                      yarl<2.0,>=1.0->aiohttp->openai==0.26.5) (2.10)
Building wheels for collected packages: openai
  Building wheel for openai (PEP 517) ... done
  Created wheel for openai: filename=openai-0.26.5-py3-none-any.whl size=67596 s                                                                                                                     ha256=ee7a7dbf798f16ab212037740dafcb034105870f5427ae8f543a4d531790f1de
  Stored in directory: /tmp/pip-ephem-wheel-cache-u7r3zif0/wheels/37/c0/19/d826e                                                                                                                     e14a79015bfd7b89bfa0674ff810b7baeb9a5a2f7d206
Successfully built openai
Installing collected packages: multidict, frozenlist, yarl, charset-normalizer,                                                                                                                      attrs, async-timeout, aiosignal, tqdm, aiohttp, openai
ERROR: pip's dependency resolver does not currently take into account all the pa                                                                                                                     ckages that are installed. This behaviour is the source of the following depende                                                                                                                     ncy conflicts.
piblo 1.2.0 requires stability-sdk@ git+https://github.com/Stability-AI/stabilit                                                                                                                     y-sdk.git, which is not installed.
Successfully installed aiohttp-3.8.4 aiosignal-1.3.1 async-timeout-4.0.2 attrs-2                                                                                                                     2.2.0 charset-normalizer-3.0.1 frozenlist-1.3.3 multidict-6.0.4 openai-0.26.5 tq                                                                                                                     dm-4.64.1 yarl-1.8.2
dave@rpipictureframe:~/pycasso $ sudo pip3 install " git+https://github.com/Stab                                                                                                                     ility-AI/stability-sdk.git"
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Collecting git+https://github.com/Stability-AI/stability-sdk.git
  Cloning https://github.com/Stability-AI/stability-sdk.git to /tmp/pip-req-buil                                                                                                                     d-9cgnd6q3
  Running command git clone -q https://github.com/Stability-AI/stability-sdk.git                                                                                                                      /tmp/pip-req-build-9cgnd6q3
  Running command git submodule update --init --recursive -q
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
    Preparing wheel metadata ... done
Collecting python-dotenv
  Using cached https://www.piwheels.org/simple/python-dotenv/python_dotenv-0.21.                                                                                                                     1-py3-none-any.whl (19 kB)
Collecting grpcio-tools==1.48.1
  Using cached https://www.piwheels.org/simple/grpcio-tools/grpcio_tools-1.48.1-                                                                                                                     cp39-cp39-linux_armv6l.whl (34.1 MB)
Collecting grpcio==1.48.1
  Using cached grpcio-1.48.1.tar.gz (22.0 MB)
    ERROR: Command errored out with exit status 1:
     command: /usr/bin/python3 -c 'import sys, setuptools, tokenize; sys.argv[0]                                                                                                                      = '"'"'/tmp/pip-install-jf6y_tca/grpcio_c2a4cba461304f8c99809eedf76863dd/setup.                                                                                                                     py'"'"'; __file__='"'"'/tmp/pip-install-jf6y_tca/grpcio_c2a4cba461304f8c99809eed                                                                                                                     f76863dd/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=                                                                                                                     f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __fi                                                                                                                     le__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/pip-pip-egg-info-gywdyhe_
         cwd: /tmp/pip-install-jf6y_tca/grpcio_c2a4cba461304f8c99809eedf76863dd/
    Complete output (13 lines):
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-install-jf6y_tca/grpcio_c2a4cba461304f8c99809eedf76863dd/se                                                                                                                     tup.py", line 263, in <module>
        if check_linker_need_libatomic():
      File "/tmp/pip-install-jf6y_tca/grpcio_c2a4cba461304f8c99809eedf76863dd/se                                                                                                                     tup.py", line 219, in check_linker_need_libatomic
        cpp_test = subprocess.Popen(
      File "/usr/lib/python3.9/subprocess.py", line 951, in __init__
        self._execute_child(args, executable, preexec_fn, close_fds,
      File "/usr/lib/python3.9/subprocess.py", line 1698, in _execute_child
        and os.path.dirname(executable)
      File "/usr/lib/python3.9/posixpath.py", line 152, in dirname
        p = os.fspath(p)
    TypeError: expected str, bytes or os.PathLike object, not list
    ----------------------------------------
WARNING: Discarding https://files.pythonhosted.org/packages/e2/13/0f91c6c7eb0d93                                                                                                                     4141743c7663e8808f4dc478dd0a7b0fd2a02a286c0d6d/grpcio-1.48.1.tar.gz#sha256=66021                                                                                                                     7eccd2943bf23ea9a36e2a292024305aec04bf747fbcff1f5032b83610e (from https://pypi.o                                                                                                                     rg/simple/grpcio/) (requires-python:>=3.6). Command errored out with exit status                                                                                                                      1: python setup.py egg_info Check the logs for full command output.
ERROR: Could not find a version that satisfies the requirement grpcio==1.48.1 (f                                                                                                                     rom stability-sdk)
ERROR: No matching distribution found for grpcio==1.48.1
dave@rpipictureframe:~/pycasso $

**Comment by BigDave365** [comment]
FWIW  after all of the above, I tried to enter an API key via setup, and it errored out with:

.bash_logout   .cache               .local     pycasso    .xsession-errors.old
.bashrc        .config              Music      Templates
dave@rpipictureframe:~ $ cd pycasso
dave@rpipictureframe:~/pycasso $ ls
examples  pycasso.service.template  requirements.txt  setup.py
images    pyproject.toml            resources         setup.sh
LICENSE   pytest.ini                scripts           src
prompts   README.md                 setup.cfg         tests
dave@rpipictureframe:~/pycasso $ bash setup.py
setup.py: line 1: from: command not found
setup.py: line 3: syntax error: unexpected end of file
dave@rpipictureframe:~/pycasso $ bash setup.sh
Traceback (most recent call last):
  File "/home/dave/pycasso/scripts/set_keys.py", line 8, in <module>
    from piblo.pycasso import Pycasso
  File "/usr/local/lib/python3.9/dist-packages/piblo/pycasso.py", line 20, in <module>
    from piblo.provider import StabilityProvider, DalleProvider, AutomaticProvider
  File "/usr/local/lib/python3.9/dist-packages/piblo/provider.py", line 15, in <module>
    import stability_sdk
ModuleNotFoundError: No module named 'stability_sdk'
dave@rpipictureframe:~/pycasso $

I manually installed stability-sdk before, so IDK what is going on.

**Comment by jezs00** [comment]
No worries. Now that you've done that try running the grpcio fix from the
installer and run pycasso.

**Comment by jezs00** [comment]
Okay I have had some problems with grpcio trying a recent reinstall so I
think that may be a culprit. This will take some time and testing, it's
possible the most recent release is not happy or playing nice with stable
diffusion, or SD has made an update that has broken something.

**Comment by BigDave365** [comment]
I tried Gripo fix, it applies, but same issue. It looks like the Stability module is not installed.

Thanks for releasing this project. It is a really great idea and I hope you can get it working on the Pi Zero W as it is a great form factor when installing into a frame.

**Comment by jezs00** [comment]
grpcio is problematic as it relies on stuff inside the kernel to work properly, which means sometimes it might be . What operating system are you guys running and is it up to date?

I just tried to reinstall completely from scratch on my 4 and it worked OK.

You could try running no requirements on installing stability-sdk and see how it goes as we don't really use keychain if we're storing things in .creds:
`sudo pip3 install stability-sdk --no-dependencies`

This might be the prompt I need to try and actually sort out a permanent solution for grpcio.

Refer to https://stackoverflow.com/questions/56357794/unable-to-install-grpcio-using-pip-install-grpcio for trying to install grpcio manually at a particular version.

**Comment by jezs00** [comment]
Well looks like I've inadvertently got keyring kind of working again while looking into this. You can also try the GRPCIO fix again once you've installed stability above with no dependencies, note that it will take a while.

**Comment by jezs00** [comment]
Hey guys, are you still experiencing issues? @BigDave365 @danmcpherson

**Comment by BigDave365** [comment]
I did/was. I think the issue might be hardware based. I was trying to install on PiZero, I started with a clean install and I cannot it to actually install and work. Many packages do not install properly and I never could get Pycasso to run , let alone produce an error log.

I eventually gave up and went back to the 3B hardware, which worked after a few tweeks (I think it did not create the prompts directory).

Long story short: with my limited skill set, the PiZero is a no-go, however 3B works and has been running the last few days.

Thanks so much for the cool project!

**Comment by jezs00** [comment]
No worries. You are right that getting it to work on a zero would be useful, so I'll leave this open and see if I can do a minimal install that uses as few libraries as possible - I still feel like stablity's dependence on grpcio is probably causing us the most headaches. I have a zero somewhere I can use to test.

**Comment by jezs00** [comment]
Cheers mate, glad it worked. I am leaving the issue open for RPI-zeroers and maybe one day I or someone else will fix it.

**Comment by jgbx35tt** [comment]
Hi @jezs00, just checking to see if there's any progress toward getting piblo installed on a RPI zero?  I've tried all the suggestions on this thread and on #51 with no success..

**Comment by jezs00** [comment]
Sorry guys no news and no developments. At present I think unless anyone else can investigate and figure something out, it's going to be a no-go on a zero. I do still think it's possible but might require quite a bit of trial and error. Please let me know if anyone has any updates.

**Comment by jezs00** [comment]
Hi all,
Big update. Looks like the issue is widespread and although I don't have a Bookworm install myself, I'm pretty sure that's the culprit. Looks like we're not doing things using best practices anymore (we probably weren't in the first place). Googling "Bookworm raspberry pi externally-managed-environment" is giving quite a few hits. Official docco supports this. The solution probably was in the error message all along:
```
╰─> To install Python packages system-wide, try apt install
    python3-xyz, where xyz is the package you are trying to
    install.
    
    If you wish to install a non-Debian-packaged Python package,
    create a virtual environment using python3 -m venv path/to/venv.
    Then use path/to/venv/bin/python and path/to/venv/bin/pip. Make
    sure you have python3-full installed.
    
    For more information visit http://rptl.io/venv
```

As such, there are a few things you guys can do to help and also to try and fix it:

- Confirm if you're on Bookworm while experiencing this issue
- **Quick and dirty fix** - Try running `  sudo pip3 install "git+https://github.com/jezs00/pycasso@$(curl -s https://api.github.com/repos/jezs00/pycasso/releases/latest | jq -r ".tag_name")" --break-system-packages
  sudo pip3 install stability-sdk @ git+https://github.com/Stability-AI/stability-sdk.git --break-system-packages
  sudo pip3 install openai @ git+https://github.com/openai/openai-python.git --break-system-packages` , let us know if this works, this is based on assumptions and I think I saw somewhere that someone succeeded using `--break-system-packages` before
- **Slower fix** - Downgrade your OS to Buster and try again, let us know if this works
- **Proper big boy fix** - See if you can create a venv according to the docs here: https://www.raspberrypi.com/documentation/computers/os.html#python-on-raspberry-pi
- If you solve it, let everyone here know so that others can benefit from the fix, and that I can maybe update the install script (I hate updating the install script)

I'll come back and mark which of the above have worked for people as time goes on.

Otherwise you can wait for me to kill my (working) frame to reinstall bookworm for the benefit of troubleshooting the issue, which, being honest, I just do not want to do right now.

**Comment by commo11** [comment]
I was able to get it up and running on a pi zero w with headless Bullseye didn't do anything special just waited forever for the script to finish running on the install.

zero@ZeroArtAI:~/pycasso $ cat /etc/os-release
PRETTY_NAME="Raspbian GNU/Linux 11 (bullseye)"
NAME="Raspbian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=raspbian
ID_LIKE=debian

zero@ZeroArtAI:~/pycasso $ sudo systemctl restart pycasso

zero@ZeroArtAI:~/pycasso $ sudo systemctl status pycasso
● pycasso.service - A system to send AI generated art to an E-Paper display through a Raspberry PI unit
     Loaded: loaded (/etc/systemd/system/pycasso.service; enabled; vendor preset: enabled)
     Active: inactive (dead) since Mon 2024-02-19 18:03:22 CST; 58s ago
    Process: 20087 ExecStart=dbus-run-session -- bash /home/zero/pycasso/scripts/run.sh (code=exited, status=0/SUCCESS)
   Main PID: 20087 (code=exited, status=0/SUCCESS)
        CPU: 28.207s

Feb 19 18:02:19 ZeroArtAI systemd[1]: Started A system to send AI generated art to an E-Paper display through a Raspberry PI unit.
Feb 19 18:02:19 ZeroArtAI dbus-run-session[20093]: /home/zero/pycasso/scripts/run.sh: line 3: gnome-keyring-daemon: command not found
Feb 19 18:03:22 ZeroArtAI systemd[1]: pycasso.service: Succeeded.
Feb 19 18:03:22 ZeroArtAI systemd[1]: pycasso.service: Consumed 28.207s CPU time.

zero@ZeroArtAI:~/pycasso $ tail pycasso.log
2024-02-19 18:02:47 DEBUG    Applying sharpness: 1.0
2024-02-19 18:02:50 DEBUG    e-Paper busy H
2024-02-19 18:02:50 DEBUG    e-Paper busy H release
2024-02-19 18:02:50 DEBUG    e-Paper busy H
2024-02-19 18:03:20 DEBUG    e-Paper busy H release
2024-02-19 18:03:20 DEBUG    e-Paper busy H
2024-02-19 18:03:20 DEBUG    e-Paper busy H release
2024-02-19 18:03:20 INFO     Send epaper to sleep
2024-02-19 18:03:20 DEBUG    spi end
2024-02-19 18:03:20 DEBUG    close 5V, Module enters 0 power consumption ...
zero@ZeroArtAI:~/pycasso $

**Comment by Rocky56gh9** [comment]
> Sorry guys no news and no developments. At present I think unless anyone else can investigate and figure something out, it's going to be a no-go on a zero. I do still think it's possible but might require quite a bit of trial and error. Please let me know if anyone has any updates.

@jezs00, I was able to get the project up and running on a Raspberry Pi Zero 2 W. It was not working on Bookworm. I started from scratch with Bullseye, 32-bit, lite. I still got a series of errors, such as the prompts directory and images directory not being created, but I was able to create them, populate the prompts, and run the project. I've attached the entire terminal readout from the setup (minus the API credentials) for reference.
[pizero install.txt](https://github.com/user-attachments/files/16110889/pizero.install.txt)

**Comment by compufreak11** [comment]
@jezs00 and @Rocky56gh9 
I just did a fresh install on a Raspberry Pi Zero 2 W w/ RaspiOS-lite Bookworm 64-bit. Confirmed working using a virtual environment as long as you install the packages manually _without_ sudo.

As you've already surmised, the issue with Bookworm is related to the new standard of python packages in virtual environments. The bash script uses sudo which attempts to install to the global package folder and you get the "externally managed environment" error.

No GRPCIO fix needed, but I did need to set the API keys manually because the bash script uses sudo again.
Just activate you virtual environment and run:
`python3 scripts/set_keys.py`
Also, I haven't tested it without the flag, but according to the omni-epd instructions, create your virtual environment with 
"--system-site-packages" to allow numpy to use global packages.
`python3 -m venv --system-site-packages your-environment-name`

**Comment by jezs00** [comment]
Looks like the bash script and readme probably needs some edits. My actual unit has died at the moment so I don't have an environment to set up and test, but if someone wants to test https://github.com/jezs00/pycasso/blob/main/setup.sh removing instances of sudo commands on bookworm, then we can probably either update the script to remove sudo, or add a flag in the installer to use sudo or not.

**Comment by jezs00** [comment]
The virtual environment issue should now be fixed. pycasso bash script now installs venv by default and activates it via run.sh. I'll leave this issue open for a month for people to play with and test if anyone's still following, and then I will close.

**Comment by jezs00** [comment]
Issue closed, presumed working as per previous comment.
